arrabbiata/src/setup.rs

```diff
@@ -168,98 +164,68 @@ where
             srs
         };
 
-        let constraints_fp: HashMap<
-            Gadget,
-            Vec<Sparse<E1::ScalarField, { MV_POLYNOMIAL_ARITY }, { MAX_DEGREE }>>,
-        > = {
-            let env: constraint::Env<E1> = constraint::Env::new();
-            let constraints = env.get_all_constraints_indexed_by_gadget();
-            constraints
-                .into_iter()
-                .map(|(k, polynomials)| {
-                    (
-                        k,
-                        polynomials
-                            .into_iter()
-                            .map(|p| Sparse::from_expr(p, Some(NUMBER_OF_COLUMNS)))
-                            .collect(),
-                    )
-                })
-                .collect()
+        // FIXME: note that the app size can be different for both curves. We
+        // suppose we have the same circuit on both curves for now.
+        let app_size = srs_size - VERIFIER_CIRCUIT_SIZE;
+
+        let verifier_fp: Verifier<E1> = Verifier::<E1> {
+            _field: std::marker::PhantomData,
         };
 
-        let constraints_fq: HashMap<
-            Gadget,
-            Vec<Sparse<E2::ScalarField, { MV_POLYNOMIAL_ARITY }, { MAX_DEGREE }>>,
-        > = {
-            let env: constraint::Env<E2> = constraint::Env::new();
-            let constraints = env.get_all_constraints_indexed_by_gadget();
-            constraints
-                .into_iter()
-                .map(|(k, polynomials)| {
-                    (
-                        k,
-                        polynomials
-                            .into_iter()
-                            .map(|p| Sparse::from_expr(p, Some(NUMBER_OF_COLUMNS)))
-                            .collect(),
-                    )
-                })
-                .collect()
+        let circuit_fp: Vec<Gadget> = {
+            let app_circuit: Vec<Gadget> = zkapp_fp.setup(app_size);
```

Comment:
Instead the zkApp defines its own size, and the setup method could:
1. return the actual circuit size.
2. could be changed into a function taking a generic zkApp (`fn setup<C: ArrabbiataCurve, App: ZkApp<C>>(app: &App) -> Vec<Gadget>`).